### ISSUE TITLE
Does api.yelp.com support Access-Control-Allow-Origin header for client-side JS?

### ISSUE THREAD

**Issue by mittonk** [Yelp/yelp-api#99]
@meydman writes in #51 :
> I am trying to create a client side call to Yelp API, but getting Access-Control-Allow-Origin error in response. I have tried various solutions, but no avail. It seems Yelp is not responding with this header. Does Yelp allow API calls from JS client?

@mittonk isn't sure.

**Comment by mittonk** [comment]
@meydman: 
1. We don't have a restriction on JS clients, but it's hard to write client-side JS that doesn't expose your secrets.
2. Mind sharing your code (without your creds)?
3. What OAuth 1.0a library are you using?

I think we may be in a similar position to
* http://stackoverflow.com/questions/16509865/url-not-allowed-by-access-control-allow-origin

Doesn't look like our old security-weak example (https://github.com/Yelp/yelp-api/pull/50/files) used the `Access-Control-Allow-Origin` header; it did use `jsonp`

**Comment by meydman** [comment]
@mittonk Thanks for looking into this. I do realize the security issue with client-side JS, which is unavoidable, that's why I was thinking that Yelp decided to block the access. Essentially the server does not respond with the CORS header. Here is the code I am using:

`var auth = {
                consumerKey: "",
                consumerSecret: "",
                accessToken: "",
                accessTokenSecret: "",
                serviceProvider : { signatureMethod : "HMAC-SHA1"}
            };

            var accessor = {
                consumerSecret: auth.consumerSecret,
                tokenSecret: auth.accessTokenSecret
            };

            var phone = '6462441149';

            var parameters = [];
            parameters.push(['phone', phone]);
            parameters.push(['oauth_consumer_key', auth.consumerKey]);
            parameters.push(['oauth_consumer_secret', auth.consumerSecret]);
            parameters.push(['oauth_token', auth.accessToken]);
            parameters.push(['oauth_signature_method', 'HMAC-SHA1']);
            parameters.push(['callback', 'cb']);

            var message = {
                'action': 'http://api.yelp.com/v2/phone_search',
                'method': 'GET',
                'parameters': parameters
            };
            console.log(message.action);

            OAuth.setTimestampAndNonce(message);
            OAuth.SignatureMethod.sign(message, accessor);

            var parameterMap = OAuth.getParameterMap(message.parameters);
            console.log(parameterMap);

            $.ajax({
                'url' : message.action,
                'data' : parameterMap,
                'dataType' : 'jsonp',
                'jsonpCallback' : 'cb',
                'async' : 'false',
                'success' : function(data, textStats, XMLHttpRequest) {
            	    console.log(data);
                }
            });

            //$.getJSON(message.action + '?', parameterMap, function (data) {
            //    console.log(data);
            //});`

As you can see, I tried JSONP as well, but without luck. I will be grateful for any help!

**Comment by watterso** [comment]
Hey there @meydman, unfortunately we do not support this header. What worked well for me in hacking together a chrome extension, I made a simple server using the [yelp python package](https://github.com/Yelp/yelp-python) and [flask](http://flask.pocoo.org/). By using a backend component you can hide your api key and control usage (and have an easy kill switch). 

I can't remember if it is required for a chrome extension but I do add CORS headers to my flask app in responding to a request (see code below) so if you _need_ to have this header, the best way to go about getting it is probably via backend component you yourself control.
```python
# NOTE: this code is flask specific
@app.after_request
def add_cross_origin(response):
  response.headers.add('Access-Control-Allow-Origin', '*')
  response.headers.add('Access-Control-Allow-Headers', 'Content-Type,Authorization')
  response.headers.add('Access-Control-Allow-Methods', 'GET')
  return response
```
**TL;DR**: No CORS is not supported by api.yelp.com

I'm going to go ahead and close this issue but if you have anymore questions, feel free to open another issue!

**Comment by mnemonicflow** [comment]
@meydman Yelp API works with [JSONP](https://en.wikipedia.org/wiki/JSONP)!

In your code the problem is with [jQuery's ajax method](https://api.jquery.com/jQuery.ajax/), which if  'dataType' is 'jsonp' it will set cache to false which means it will append an additional "_" (underscore) parameter to the query string.

This is what the docs says about the cache parameter:
```
cache (default: true, false for dataType 'script' and 'jsonp')
Type: Boolean
If set to false, it will force requested pages not to be cached by the browser. Note: Setting cache to false will only work correctly with HEAD and GET requests. It works by appending "_={timestamp}" to the GET parameters. The parameter is not needed for other types of requests, except in IE8 when a POST is made to a URL that has already been requested by a GET.
```

Because your request is signed with OAuth v1.0a *before* sending it to jQuery's ajax method, that parameter is **NOT** part of the signing process.

And when the HTTP request is made by jQuery's ajax method to the Yelp API it will contain the additional underscore parameter in the query string which will result in the following error: `{"error": {"text": "Signature was invalid", "id": "INVALID_SIGNATURE", "description": "Invalid signature. Expected signature base string: ..."}`
This basically means that the server side the signature verification fails.

If you look carefully at the "Expected signature base string" you will notice an additional `_%3D1464707738290` parameter besides those you passed to  `var parameterMap = OAuth.getParameterMap(message.parameters);`


To make it work using jQuery's ajax method you have to prevent jQuery from adding the additional underscore by passing 'cache': true when making the jsonp request:
```javascript
$.ajax({
            'url' : message.action,
            'data' : parameterMap,
            'dataType' : 'jsonp',
            'jsonpCallback' : 'cb',
            'async' : 'false',
            'cache': true,
            'success' : function(data, textStats, XMLHttpRequest) {
                console.log(data);
            }
        });
```
If you use Angular.js you can also see this example: https://stackoverflow.com/questions/23716264/yelp-api-and-angularjs/26152065#26152065 , which I've tested and it also works with Yelp API using JSONP.

**Comment by TechhDan** [comment]
Hi,
Does anyone have a working sample of this? I've been trying but no luck.

`$.ajax({
    url: url,
    beforeSend: function(xhr) {
      xhr.setRequestHeader( "Authorization", "Bearer " + api_key);
    },
    method: 'GET',
    dataType: 'jsonp',
    jsonpCallback: 'cb',
    async: false,
    cache: true,
  }).done(function(result) {
    console.log(result)
  });`
